5-data-modeling/event-manifest-populator/src/main/scala/com/snowplowanalytics/snowplow/eventpopulator/Utils.scala

```diff
@@ -0,0 +1,216 @@
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.eventpopulator
+
+// scalaz
+import scalaz.{Ordering => _, _}
+import Scalaz._
+
+// Scala
+import scala.util.control.NonFatal
+import scala.collection.convert.decorateAsScala._
+
+// AWS SDK
+import com.amazonaws.auth.DefaultAWSCredentialsProviderChain
+import com.amazonaws.services.s3.AmazonS3ClientBuilder
+import com.amazonaws.services.s3.model.ListObjectsRequest
+
+// Java
+import java.nio.charset.StandardCharsets.UTF_8
+
+// Jackson
+import com.fasterxml.jackson.databind.{ ObjectMapper, JsonNode }
+
+// Iglu client
+import com.snowplowanalytics.iglu.client.Validated
+import com.snowplowanalytics.iglu.client.validation.ProcessingMessageMethods._
+
+// Apache Commons Codec
+import org.apache.commons.codec.binary.Base64
+
+// Joda time
+import org.joda.time.DateTime
+import org.joda.time.format.DateTimeFormat
+import org.joda.time.format.DateTimeFormatterBuilder
+
+// This project
+import Main.JobConf
+
+/**
+  * Util functions and constant that not directly related to spark job
+  */
+object Utils {
+
+  // Column indexes in TSV
+  val EtlTimestampIndex = 2
+  val EventIdIndex = 6
+  val FingerprintIndex = 129
+
+  /**
+    * Date format in archive buckets: run=YYYY-MM-dd-HH-mm-ss
+    * In EmrEtlRunner: "%Y-%m-%d-%H-%M-%S"
+    */
+  val runIdFormat = DateTimeFormat.forPattern("YYYY-MM-dd-HH-mm-ss")
+
+  /**
+    * Shortened `runIdFormat` that can be used for manual input
+    */
+  val dayFormat = DateTimeFormat.forPattern("YYYY-MM-dd")
+
+  /**
+    * Combined `dayFormat` and `runIdFormat` that can be used
+    * both for precise and shorten manual input
+    */
+  val inputFormatter = new DateTimeFormatterBuilder()
+    .append(null, List(runIdFormat.getParser, dayFormat.getParser).toArray)
+    .toFormatter
+
+  implicit def dateTimeOrdering: Ordering[DateTime] =
+    Ordering.fromLessThan(_ isBefore _)
+
+  private val UrlSafeBase64 = new Base64(true) // true means "url safe"
+
+  private lazy val Mapper = new ObjectMapper
+
+  /**
+    * Data extracted from EnrichedEvent and storing in DynamoDB
+    */
+  case class DeduplicationTriple(eventId: String, fingerprint: String, etlTstamp: String)
+
+  /**
+    * Error happening when TSV doesn't match expected number of lines
+    *
+    * @param columns encountered number of columns
+    * @param line original event TSV line
+    */
+  case class ParsingError(columns: Int, line: String) extends RuntimeException {
+    def message: String = s"Wrong EnrichedEvent TSV columns number. Expected 130, got [$columns]. Line: [$line]"
+  }
+
+  /**
+    * Get run ids from `jobConf.enrichedInBucket` happened since `jobConf.since`
+    *
+    * @param jobConf input configuration
+    * @return list of timestamps (run ids)
+    */
+  def getRuns(jobConf: JobConf): List[String] = {
```

Comment:
@alexanderdean I think I found a bug! How many runs were in the bucket you tested it in? S3 truncates keys to 1000 in list requests. If there were >1000 then it probably didn't reach specified date at all.